specs/agents/tracing-spans-service-target.md

```diff
@@ -0,0 +1,208 @@
+|-------------------------------------------------|---------|---------------|-------------|---------------------|
+| `span.type`                                     | `db`    | `db`          | `messaging` | `messaging`         |
+| `span.subtype`                                  | `mysql` | `mysql`       | `rabbitmq`  | `rabbitmq`          |
+| `span.context.service.target.type`              | `mysql` | `mysql`       | `rabbitmq`  | `rabbitmq`          |
+| `span.context.service.target.name` (1)          |         | `my-db`       |             | `my-queue`          |
+| `span.context.destination.service.resource` (2) | `mysql` | `mysql/my-db` | `rabbitmq`  | `rabbitmq/my-queue` |
+
+(1) Value depends on the instrumented backend, see [below](#field-values) for details.
+
+(2) Value is inferred and sent by APM agents in Phase 2, inferred on APM Server once Phase 3 is complete
+
+## Implementation details
+
+### Field values
+
+This specification assumes that values for `span.type` and `span.subtype` fit the [span_types.json](../../tests/agents/json-specs/span_types.json) specification.
+
+- `span.context.service.target.*` fields should be omitted for non-exit spans.
+- Values set by user through the agent API should have priority over inferred values.
+- `span.context.service.target.type` should default to the same value as `span.type`.
+- When a `span.subtype` value is available, `span.context.service.target.type` should have the same value as `span.subtype`.
+
+On agents, the following algorithm should be used to infer the values for `span.context.service.target.*` fields. 
+```javascript
+// span created on agent
+span = {};
+
+if (span.isExit) {
+  service_target = span.context.service.target;
+  
+  if (!service_target.type) { // infer type from span type & subtype
+      
+    // use sub-type if provided, fallback on type othewise
+    service_target.type = span.subtype || span.type;
+  }
+  
+  if (!service_target.name) { // infer name from span attributes
+      
+    if (span.context.db.instance) {  // database spans
+      service_target.name = span.context.db.instance;
+
+    } else if (span.context.message) { // messaging spans
+      service_target.name = span.context.message.queue.name
+
+    } else if (context.http.url) { // http spans
+      service_target.name = context.http.host;
+      if (context.http.url.port > 0) {
+        service_target.name += ":" + context.http.port;
+      }
+    }
+
+  }
+} else {
+    // non-exit spans should not have service.target.* fields
+    span.context.service.target = undefined;
+}
+```
+
+The values for `span.context.db.instance` are described in [SQL Databases](./tracing-instrumentation-db.md#sql-databases).
+
+The values for `span.context.message.queue.name` are described in [Messaging context fields](./tracing-instrumentation-messaging.md#context-fields)
+
+### Phase 3
+
+In Phase 3 the `span.destination.service.resource` field value is inferred on APM Server with the following algorithm:
```

Comment:
Is the goal here to always populate `span.destination.service.resource` and `span.service.target.*` fields? So basically, for older apm agents, derive `span.service.target.*` from  `span.destination.service.resource` and for backwards data compatibility derive `span.destination.service.resource` from  `span.service.target.*` when sent?

If so, is the UI also implementing the change to use `span.service.target.*` or are they still dependend on the currently populated field  `span.destination.service.resource` ? I am wondering if we really need both.

Reply:
The UI won´t change and still relies on `span.destination.service.resource`, this is why we have to keep it compatible.

The main reason was complexity: making UI able to support both while keeping the map & service dependencies consistent would very likely have become a maintenance burden. Also, the "maps v2" ongoing project might help changing this later, with this change at least we provide structured data we can rely on.